src/Microsoft.ML.Transforms/Text/TextCatalog.cs

```diff
@@ -362,6 +362,31 @@ public static class TextCatalog
             => new WordBagEstimator(Contracts.CheckRef(catalog, nameof(catalog)).GetEnvironment(),
                 outputColumnName, inputColumnName, ngramLength, skipLength, useAllLengths, maximumNgramsCount, weighting);
 
+        /// <summary>
+        /// Create a <see cref="WordBagEstimator"/>, which maps the column specified in <paramref name="inputColumnName"/>
+        /// to a vector of n-gram counts in a new column named <paramref name="outputColumnName"/>.
+        /// </summary>
+        /// <remarks>
+        /// <see cref="WordBagEstimator"/> is different from <see cref="NgramExtractingEstimator"/> in that the former
+        /// tokenizes text internally and the latter takes tokenized text as input.
+        /// </remarks>
+        /// <param name="catalog">The transform's catalog.</param>
+        /// <param name="outputColumnName">Name of the column resulting from the transformation of <paramref name="inputColumnName"/>.
+        /// This column's data type will be known-size vector of <see cref="System.Single"/>.</param>
+        /// <param name="inputColumnName">Name of the column to take the data from.
+        /// <param name="maximumNgramsCount">Maximum number of n-grams to store in the dictionary.</param>
+        /// <param name="termSeparator">Separator used to separate terms/frequency pairs.</param>
+        /// <param name="freqSeparator">Separator used to separate terms from their frequency.</param>
+        /// This estimator operates over vector of text.</param>
+        public static WordBagEstimator ProduceWordBagsPreDefinedWeight(this TransformsCatalog.TextTransforms catalog,
```

Comment:
@ericstj @luisquintanilla @tannergooding thoughts on this api name? The normal one is just called `ProduceWordBags`, but when naming this one that it can potentially cause ambiguity issues due to the default parameters not needing to be specified.

I had 2 options, either name it something different, or not have default parameters and make the users specify the term separator and frequency separator manually each time. I went with the first approach, but want to hear your opinions on it.

Reply:
@tarekgh your thoughts as well if you have time.

Reply:
My thoughts:

In main scenarios is it expected to be called once or twice in the code? If yes, then I prefer using `ProduceWordBags` without any optional parameter. Will be just overload. If the answer is no, then using a different name would be better, I guess. We may try to suggest a better name than `ProduceWordBagsPreDefinedWeight`. May be ``ProduceWordBagsEstimator` is simpler?

Reply:
With the current name, I think a "With" should be in there (`ProduceWordBagsWithPreDefinedWeight`).  Though since the caller doesn't get to specify it "PreDefined" seems more like "Default".

I assume the real reason a caller wants this method is to specify the term and/or freq separators.  Can this not just be added as a true extra optional parameter overload using the established patterns?

```diff
+       [EditorBrowsable(EditorBrowsableState.Never)]
        public static WordBagEstimator ProduceWordBags(this TransformsCatalog.TextTransforms catalog,
            string outputColumnName,
            string inputColumnName = null,
-           int ngramLength = NgramExtractingEstimator.Defaults.NgramLength,
+           int ngramLength,
-           int skipLength = NgramExtractingEstimator.Defaults.SkipLength,
+           int skipLength,
-           bool useAllLengths = NgramExtractingEstimator.Defaults.UseAllLengths,
+           bool useAllLengths,
-           int maximumNgramsCount = NgramExtractingEstimator.Defaults.MaximumNgramsCount,
+           int maximumNgramsCount,
-           NgramExtractingEstimator.WeightingCriteria weighting = NgramExtractingEstimator.WeightingCriteria.Tf)
+           NgramExtractingEstimator.WeightingCriteria)
            => new WordBagEstimator(Contracts.CheckRef(catalog, nameof(catalog)).GetEnvironment(),
                outputColumnName, inputColumnName, ngramLength, skipLength, useAllLengths, maximumNgramsCount, weighting);

+       public static WordBagEstimator ProduceWordBags(this TransformsCatalog.TextTransforms catalog,
+           string outputColumnName,
+           string inputColumnName = null,
+           int ngramLength = NgramExtractingEstimator.Defaults.NgramLength,
+           int skipLength = NgramExtractingEstimator.Defaults.SkipLength,
+           bool useAllLengths = NgramExtractingEstimator.Defaults.UseAllLengths,
+           int maximumNgramsCount = NgramExtractingEstimator.Defaults.MaximumNgramsCount,
+           NgramExtractingEstimator.WeightingCriteria weighting = NgramExtractingEstimator.WeightingCriteria.Tf,
+           char termSeparator = ';',
+           char freqSeparator = ':')
+           => new WordBagEstimator(parameters go here);
```

Reply:
Which overload will be chosen if the caller is passing only the first parameter?

Reply:
It could be added as a true optional parameter if the user had to specify it each time. Which as you said @tarekgh it will be used probably 1 time only for each pipeline, so thats probably not a big deal.

Would changing the existing method though by adding new default parameters count as a breaking change @ericstj?

If it is a breaking change I could just change the name I gave it to match the existing names and just make those parameters required each time.

Reply:
> Which overload will be chosen if the caller is passing only the first parameter?

If the compiled in the past, the old one.  If they compile after, the new one.  (The old method, now marked as [EB(Never)] is only called by a) legacy callers or b) someone who happened to specify all of those parameters and none of the new ones... the compiler will still (and only) call it if the callsite has an exact match to the signature.

Since both methods are creating a `WordBagEstimator` I'm assuming that "specifying the weighting" and "specifying the freq/term-separator" aren't actually conflicting options.  If they _are_ conflicting options, then some sort of differentiated name is warranted.

Reply:
> Would changing the existing method though by adding new default parameters count as a breaking change @ericstj?

Adding the parameters to the existing method is a breaking change (a really, really bad one).  Adding it the way I suggested is the way to make it look like all you did was add new default parameters.

From Framework Design Guidelines, 3rd edition, sec 5.1 (General Member Design Guidelines):

> DO move all default parameters to the new, longer overload when adding optional parameters to an existing method.
>
> [prose that says to do what I said to do above]

and down in Appendix D (breaking changes) (emphasis mine):

> D.11.2 Adding or Removing a Method Parameter
>
> Methods in the .NET CLR are identified by their signature, which is composed
of the name, return type, and ordered list of parameters. Removing
a parameter, adding a required parameter, _**or adding an optional parameter**_
all count as changing the signature and, therefore, are logically the
same as deleting the original method. See section D.9.3 for information on
the runtime impact of removing a member.

Reply:
@bartonjs the separators and the weight aren't conflicting options. The default should be just to use the frequency of the words, but there isn't a problem if they are both specified together (though I would bet that most people don't want to specify both together and probably would give them unexpected, though not incorrect, behavior if they do that).

I think making the name the same with a new overload where the user has to specify the separators is probably the best based on this convo. We don't want the separators specified unless the user really needs them, since thats the flag that tells word bag that it needs to handle the input data differently.

Reply:
@bartonjs I have made the changes.

Reply:
That new catalog overload looks a lot nicer to me :smile: